### PR TITLE
[RFC] Add a general graphic display API

### PR DESCRIPTION
**Pull request by diantox**
**Goal**
To implement a general graphic display API.

**Rationale**
All graphic displays implement the same fundamental functionality: change the color of a pixel or group of pixels. Different displays go about this differently, some displays (like [Sharp Memory LCDs](https://www.sharpsma.com/products?sharpCategory=Memory%20LCD)) can only update entire lines long the horizontal axis, while other displays like those based on the [SSD1351](http://www.solomon-systech.com/en/product/display-ic/oled-driver-controller/ssd1351/) can update partial lines down to a single pixel, or even change scan direction (scan vertically instead of horizontally). Atop this, different developers expose the same functionality under different API surfaces (ie. `void update_pixel(uint8_t x, uint8_t y, bool color)` vs `int update_px(struct px)`).

An application that uses a particular display driver with an arbitrary API surface can't change displays without a moderate to substantial rewrite of it's display subsystem. With a general display API, that same application could change displays with very little rewriting.

**What This Should Cover**
- width
- height
- color space
- back/frontlight brightness (particularly useful for OLEDs where each pixel _IS_ the backlight)
- orientation (if the display is _physically_ mounted at an angle)
- sleeping
- waking
- various drawing operations (ie. drawing pixels, lines, frames)
- display specific information and operations

**What This Shouldn't Cover**
- framebuffering
- color conversion (ie. converting colors from 18-bit to 16-bit or 1-bit)
- polyfilling unsupported operations (ie. polyfilling a Sharp Memory LCD to update a single pixel)

**My Proposal**
My pull request includes the SEGDL API (Simple Embedded Graphic Display Library) that covers all of the above functionality, a [Sharp LS013B7DH05](https://www.sharpsma.com/products?sharpCategory=Memory%20LCD&sharpProductRecordId=1725803) driver that implements the SEGDL API, and sample code for that driver.

I am currently working on a framebuffer that will sit on top of a SEGDL API implementation. The framebuffer will cover polyfilling unsupported operations, but not color conversion.

### CODE REVIEW

**Comment by carlescufi** [comment]
@M1ras we don't use merge commits in order to rebase to the latest. Please use `git rebase upstream/master` or similar and get rid of the merge commit. Thanks!

**Comment by vanwinkeljan** [comment]
Note that a low level display API is already available in zephyr, see [include/display.h](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/display.h)

File: include/display/segdl.h [hunk 1]
```diff
@@ -0,0 +1,974 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @author Miras Absar <miras@miras.io>
+ * @brief The SEGDL API.
+ *
+ * The SEGDL (Simple/Shitty Embedded Graphic Display Library) API is a
```
Comment (carlescufi) [inline include/display/segdl.h]: Is this original work for Zephyr, or does this already exist elsewhere?

**Review by vanwinkeljan — CHANGES_REQUESTED** [review]

File: include/display/segdl.h [hunk 2]
```diff
@@ -0,0 +1,974 @@
...
+ * cover general display information and functionality.
+ * It covers...
+ *
+ * - width
+ * - height
+ * - color space
+ * - max brightness
+ * - orientation
+ * - drawing pixels
+ * - drawing lines
```
[the remark this answers is not in the document]
Comment (M1ras) [reply include/display/segdl.h]: Sharp Memory LCDs can only draw entire horizontal lines at a time (they can't update single pixels), while more complicated displays like ones based on the SSD1351 can update single pixels, partial lines (lines that don't cover the entire width or height of the display), and entire lines (lines that do cover the entire width or height of the display). This wasn't meant to be a GUI API, it was meant to expose a display's base functionality (`struct segdl_api` allows developers to signal what functions are and aren't supported by a display).

**Comment by M1ras** [comment]
@carlescufi @vanwinkeljan I didn't realize Zephyr already had a display API. I think it makes sense to close this PR as DNM. I'll open a new PR with just the Sharp LS013B7DH05 display driver. I have a few concerns about limitations in the current display API; I'll open an issue about that.

**Comment by vanwinkeljan** [comment]
@M1ras Feel free to propose improvements to the API via PR. One that I see useful here is to extend the display capabilities struct with min x & y pixel parameter. In the case of the sharp display we could set min x = display width & min y = 1.

**Comment by carlescufi** [comment]
> @carlescufi @vanwinkeljan I didn't realize Zephyr already had a display API. I think it makes sense to close this PR as DNM. I'll open a new PR with just the Sharp LS013B7DH05 display driver. I have a few concerns about limitations in the current display API; I'll open an issue about that.

As @vanwinkeljan mentions, please open a PR for the driver and an additional one to improve the current display API. They could strictly also be in the same PR if the driver requires the additional APIs you might add.

**Comment by jfischer-phytec-iot** [comment]
@carlescufi @vanwinkeljan I suggest to close it.

**Comment by carlescufi** [comment]
@M1ras do you have plans to follow-up on this PR? Otherwise we will close it

**Comment by carlescufi** [comment]
Closing as stale

**Pull request closed without merging**